### ISSUE TITLE
Recognize Opening Hours with Machine Learning from Photos

### ISSUE THREAD

**Issue by l-salewski** [streetcomplete/StreetComplete#4775]
**Use case**

I found that entering opening hours by hand is quite tedious, which is a shame since it is really useful data. Instead it would be nice if one could snap a picture of the opening hours and the app runs an OCR algorithm on it. Based on that outcome it suggests data, which the user then sanity checks and confirms.

**Proposed Solution**

1. Take a picture of the opening times signs found at many stores. The user should be instructured to have the sign reasonable in the frame and not to break any privacy related regulations in their location.
2. Run an OCR algorithm on it (e.g. [Tessaract](https://github.com/tesseract-ocr/tesseract), ideally on the device of the user). This results in some text with x,y coordinates, which we simply flatten into lines.
3. Delete the image. We do not need it anymore.
4. Feed the flattened output raw text to a Language Model to predict the value for the ``opening_hours`` key. Ideally this should also run on the device of the user. The below example shows this with ChatGPT, but it should be fairly simple to do this with a smaller language model and include few-shot examples for in-context learning. Providing a few examples enables smaller language models to perform better and also increases the quality of the suggestions. These examples could be created based on the [wiki examples](https://wiki.openstreetmap.org/wiki/DE:Key:opening%20hours?uselang=de).
![grafik](https://user-images.githubusercontent.com/71447327/215333592-f2a3e2c4-71f6-42f8-bda9-f23830e519f7.png)
5. Show the parsed result in the existing UI of the app.
6. If the user thinks everything is correct, they submit the data.

Overall this solution would not need any large data collection, model training and would be privacy persevering. Any errors from the models will be corrected by the user, as this approach is solely aimed to reduce the friction in entering the data in the app. The users still have to check it. One downside would be that even the quantized machine learning models (for OCR and the Language Model for parsing) can have sizes in the hundreds of megabytes but we could make this an optional in-app download for users who may want to use this feature.
On the plus side, this feature could reduce the time need to enter opening hours from a ~1 minute to simply snapping an image and pressing the okay button (if everything is correct).

Unfortunately, I am not very familiar with Android app development. In python it should be pretty easy, maybe 100 LOC of tesseract and pytorch, but that probably wont help for StreetComplete. Of course I would be open to help on the machine learning side and it seems there are a number of [Pytorch Tutorials for Android](https://pytorch.org/mobile/android/).

**Comment by eginhard** [comment]
Duplicate of #1186

**Comment by l-salewski** [comment]
I searched for similar issues but apparently my searches used the wrong keywords. Sorry for that.

I think the topics discussed in these issues are completely valid (e.g. size of ML models), however the proposed solution to use a language model for parsing the OCR result is novel.

It actually seems to adress some of the [issues (e.g. non-trivial cases, different languages)](https://github.com/streetcomplete/StreetComplete/issues/1186#issuecomment-756036717) mentioned in the other thread (this is without any few-shot examples, so it might work even better once 10-20 examples are prefixed).

![grafik](https://user-images.githubusercontent.com/71447327/215340883-c2149dac-4f39-43fc-a47b-1d3ad96026ef.png)

**Comment by westnordost** [comment]
On Android one can use any Java library, or more general, any JVM library, e.g. it could be written in Kotlin, Scala or Groovy too. Also, it is possible to use native libraries (written in C++ etc) in an Android project through Java Native Interface (JNI) as long as those bindings are provided. (See also https://developer.android.com/ndk/ )

So, the OCR+ML model to map the information in the photo to OSM tags do not need to have anything to do with Android, or Java for that matter.

If you are well versed in the matter, I would like to suggest that you work on this in a separate public repository and produce a (Java) library of it so that it can be used not only in StreetComplete, but also Vespucci, EveryDoor or even desktop editors like JOSM etc.

One thing I would very much like to avoid is to have this feature based on a web service that would require sending the pictures to that web service every time because StreetComplete should work offline, too. And this constraint somewhat excludes ChatGPT AFAIK.

**Comment by mnalis** [comment]
Related: https://mastodon.online/@everydoor@en.osm.town/113455326650743188 (_"Ian's LLM photo-to-tags bridge being integrated into Every Door"_)

**Comment by mnalis** [comment]
Thanks for the links, @mcliquid ! I was not aware of the details till now.

I should also note that while it is certainly much easier (and less resource intensive, not to mention multiplatform by design) to just call an web API instead of having engine + training database locally on the phone; one should take note that there are (serious?) **privacy (and publicity) issues with uploading photos to OpenAI's servers** (and in addition, one needs to be online / have big data plan).

**Comment by digitalcraftsman** [comment]
/ping @westnordost 

> One thing I would very much like to avoid is to have this feature based on a web service that would require sending the pictures to that web service every time because StreetComplete should work offline, too. And this constraint somewhat excludes ChatGPT AFAIK.

I sympathise with the goal to keep StreetComplete usable while being offline. **Though, entering opening hours is relatively elaborative.**

Making the usage of ChatGPT (via an api call to their api) **optional**, would not undermine your concerns while streamling the process for users willing to use ChatGPT. 

## Concept 

Entering opening hours could be boiled down to:

1. taking a picture,
2. verifying ChatGPtT's response by the user.

 <details>
  <summary>Example prompt</summary>
  
> The following wiki article describes how to encode opening hours for points of interest in Openstreetmap: https://wiki.openstreetmap.org/wiki/Key:opening_hours/specification. Please encode the opening hours in the attached picture and return only the encoded version without further text or explanatio

</details> 

I've tested the above prompt with photos of signs for opening hours taken from weird angles, with reflections, photos containing the sign and unrelated content, hand written text etc. **So far ChatGPT did an excellent job.**

## Offline-only workflow

While more complicated, the workflow could adjusted as follows:

1. users take photo of opening of a point for interest
1. once able access the internet, the photos are processed by ChatGPT in the background
1. the user verifies ChatGPT's response
1. all human-verified POIs are uploaded to OSM

## Privacy concerns

 @mnalis's privacy concerns could be mitigated by stripping EXIF metadata from photos before sending them to OpenAI. As an optional feature users can evaluate potential privacy risks for themselves.

**Comment by matkoniecz** [comment]
> Making the usage of ChatGPT (via an api call to their api) optional, would not undermine your concerns while streamling the process for users willing to use ChatGPT.

for start, who would pay for that?

**Comment by digitalcraftsman** [comment]
Bad news 🙁

The [free tier of ChatGPT](https://chatgpt.com/#pricing) lists access to GPT-4o mini. Earlier I've tested it with 3 photo uploads.

The free tier features "Limited access to file uploads". Digging further, the [ChatGPT docs](https://help.openai.com/en/articles/8555545-file-uploads-faq#h_52e5a07158) state:

> Please note that users are be able to upload up to 80 files every 3 hours on GPT-4o, and **Free users can upload up to 3 files per day**. We may reduce the limit during peak hours to keep GPT-4o accessible to the widest number of people.

Earlier I stayed just under the upload limit of 3 photos. Now, uploads aborts with an error. I expected a limit, but hoped for a higher one.

It was intended, that the used services / libraries do not induce any costs and provide reasonable limits to keep such a feature usable in the StreetComplete app.

Sorry for the disturbance due to my previous ethusasism for such a feature.

**Comment by l-salewski** [comment]
Whilst external API-based implementations might be easier, they come with many of the above discussed downsides of cost, privacy and difficult offline workflows.
But Android actually already has a [local LLM Inference API](https://ai.google.dev/edge/mediapipe/solutions/genai/llm_inference/android), which allows an app to do the above mentioned tasks entirely on device, without needing to implement the neural network inference code. I think this would be a much cleaner and easier way to approach this problem.
In my experience the availabe Gemini models are actually quite strong and a few few-shot examples would probably enable the model to suggest the opening hours with high accuracy.

**Comment by matkoniecz** [comment]
> [local LLM Inference API](https://ai.google.dev/edge/mediapipe/solutions/genai/llm_inference/android)

is it actually LLM (so input and output is limited to text) or is it also including high-quality image processing and OCR?

from

> The LLM Inference API lets you run large language models (LLMs) completely on-device for Android applications, which you can use to perform a wide range of tasks, such as generating text, retrieving information in natural language form, and summarizing documents. The task provides built-in support for multiple text-to-text large language models, so you can apply the latest on-device generative AI models to your Android apps.

it does not seem to support image processing

**Comment by l-salewski** [comment]
The current limitation of the LLM API only being able to process text could be worked around with utilizing an OCR package like Google's tesseract. However, given that Google's latest models are all multi-modal, it is fair to assume that the API will soon be extended to multi-modal inputs.
CLIP unfortunately would not help in this setting, as it is a contrastive model that can tell us whether an image and a text match (or not), but we would need a generative model that can parse the OCR result/image and output the data in a structured way.

**Comment by mnalis** [comment]
Just to note that the **issue is not only technical, but also political** as the opinions regarding AI are heavily charged (i.e. one should take into consideration that there seems to be huge anti-AI sentiment, so any decision to employ AI is quite possibly going to have negative public relations fallout). 

See e.g. comment sections on any AI news story, but also recent [anti-AI rage against JOSM maintainer](https://community.openstreetmap.org/t/ai-ferrets-take-over-josm/125426) (especially initial non-edited versions were pure venom spewing). While certainly some AI uses will have it worse than others, I'm no longer sure that _any_ AI usage is likely to go smoothly. 

In any case, if it becomes technically feasible, and is considered for implementation, I'd recommend making a **completely separate app** for that, with which SC might communicate using intents or something (e.g. like StreetMeasure) (that might also be required due to less-free ToS of certain APIs, but even if there are no restrictions and APIs are fully open source, it would still be benefitial to separate it from core StreetComplete app for reasons mentioned above).

**Comment by mnalis** [comment]
ref. related EveryDoor attempts where AI turned out to be more problematic then helpful:
- https://github.com/Zverik/every_door/issues/531
- https://github.com/Zverik/every_door/pull/894

**Comment by DonnerWolfBach** [comment]
I would be interested in this too.
@mnalis 
> In any case, if it becomes technically feasible, and is considered for implementation, I'd recommend making a completely separate app for that, with which SC might communicate using intents or something (e.g. like StreetMeasure) (that might also be required due to less-free ToS of certain APIs, but even if there are no restrictions and APIs are fully open source, it would still be benefitial to separate it from core StreetComplete app for reasons mentioned above).

How easy would it be to get the "hey we need to get some opening hours in this area" information via an api or something into an external app? Bc for me this looks the most difficult part.

> for start, who would pay for that? ChatGPT API is not free last time I checked*, and "put you chatgpt API key" is something that will be not doable by 99% of users

I would for mine.

**Comment by mnalis** [comment]
> How easy would it be to get the "hey we need to get some opening hours in this area" information via an api or something into an external app? Bc for me this looks the most difficult part.

As noted, we already have external app ([StreetMeasure](https://play.google.com/store/apps/details?id=de.westnordost.streetmeasure)) which  passes such questions/answers via API between StreetComplete and StreetMeasure apps. So that part is as easy as copy/paste.

Most everything else is problematic, though.

**Comment by mmann1123** [comment]
Hey all, i just posted a method that is working pretty well using ML Kit text recognition. I am just testing it now, I am sure there's issues I am not considering. See #6679 

https://github.com/user-attachments/assets/ce0adc68-086a-4c64-ac17-94a5cb4e3489

**Comment by westnordost** [comment]
https://github.com/streetcomplete/StreetComplete/issues/1186#issuecomment-3709210006

**Comment by pktiuk** [comment]
I think this issue is worth reopening and reevaluating.    
Hardware and software landscape changed a lot.
I think it could be implemented either via classical OCR+LLM(for making sense out of text and formatting it into OSM format) or Multimodal.

Nowadays more and more devices come with NPU units which could handle this kind of load locally.     
Moreover there are more and more integrations and tools to handle this kind of task.     
Some potential tooling examples:

- ML Kit with new [ GenAI Image Description API]( https://developers.google.com/ml-kit/genai/image-description/android)
- Local llms on Android [GitHub page](https://github.com/dineshsoudagar/local-llms-on-android) (LiteRT + ONNX)
- [llama.cpp](https://github.com/ggml-org/llama.cpp) LLM inference engine which could be used locally
- MLC LLM: Universal LLM Deployment Engine With ML Compilation https://llm.mlc.ai

**Comment by westnordost** [comment]
We'd need:

- device-local OCR in Kotlin Multiplatform (so that it works on iOS, Android, ...)

- local model that knows how to convert a given blob of text into an OSM opening hours string

-  local engine that runs this model that works in kotlin multiplatform

Thrice of which could and should be separate projects and be used as dependencies.

**Comment by pktiuk** [comment]
@westnordost yes, and these libraries already exist. Developers just need to decide which ones are the most suitable.

> device-local OCR in Kotlin Multiplatform (so that it works on iOS, Android, ...)

https://github.com/ismoy/ImagePickerKMP -library

> local model that knows how to convert a given blob of text into an OSM opening hours string

Local model as library https://llm.mlc.ai/ or ONNX

**Comment by westnordost** [comment]
Lol, well, if you think so, you are welcome to implement it.

**Comment by pktiuk** [comment]
@westnordost 

> Lol, well

Did I make any mistake? I listed some libraries and you did just claimed that libraries are needed.

> you are welcome to implement it

I would prefer to start by reopening this thread and discussion, and then collectively decide how it should work and how it should be implemented.

> you are welcome to implement it

And then what? I will prepare PR and you will just claim that you don't like it because it did not fulfill requirements not listed anywhere.

--------------------

To be honest I am really disappointed with your response. It seems that you are not very interested in working with community. 🫤 That's a shame, because I really like the idea behind StreetComplete.

**Comment by mnalis** [comment]
> To be honest I am really disappointed with your response. 

**TL;DR:** Note that the issue has been discussed ad nauseam already, and so far nobody has even produced **convincing evidence** (beyond referring on all the hype) that it _might work_ **reliably enough**[^3] -- not even with cloud models, much less local ones which are _significantly_ weaker. 

So doing that (which does not require any programming experience, just some time) should be a very first step before all the effort at any implementation should even be considered.

---

In any case, I would think any such potential future implementation of LLM/OCR to read opening hours should _**definitely**_ be a **separate app**[^1], i.e. like we have [StreetMeasure](https://github.com/streetcomplete/StreetMeasure) companion-app for AR-measurements. 

I would suggest someone interested in that[^2] feature should:
- create a collection of at least a 100 random images of various types of opening hours images preferably in various countries (both simpler and more complex) - e.g. from personal collection or Panoramax, Mapillary etc. (and store them on their own GitHub repo or somewhere, so it can be reused by others)
- run them on the cloud LLMs of their choice to turn into `opening_hours` format
- report how many of them were fully[^3] correct.

I that gives **great** results, it might become interesting for someone to write such a separate app using a local model (e.g. webapp should be relatively easy and crossplatform approach) so wider range of users could test it, and **only then** (when such app gets good testing with promising results) it should be considered to add such _"StreetHoursAI"_ app as helper for StreetComplete (like we already do with StreetMeasure), resulting in opening the issue in this repo.

I've not done any serious testings, but the few examples I've run do not give me confidence that it would be useful (while I used cloud, note I did not use any payed models). Still, all the free cloud LLMs are way ahead of any local LLM implementation AFAIK. 

You can try one such implementation on https://image-to-osm.vercel.app/ - upload your samples there and report the results as noted above.


[^1]: it does not make sense to significantly bloat the size of the whole app for the potential benefit of the single Quest, not to mention other potential issues like licensing, models etc.
[^2]: even those those who cannot do the whole coding thing themselves as the are not programmers (and can't get LLMs to write code for them). E.g. you yourself can do that if you are interested in helping!
[^3]: as that is the interesting case. If the result requires used to fix the data, it is likely to be more time confusing than just entering data from scratch as well as much more likely to enter invalid data in OSM as they did not check the result carefully. We _already_ had such issues with some people accepting AR measurements even when the results are obviously incorrect and they've explicitly warned about it.

**Comment by westnordost** [comment]
So, first of all, Pawel, it took me about a minute to discover that ImagePickerKMP has nothing to do with OCR and that MLC LLM is not KMP and also is at least marketed to be for LLMs, not for any (local) ML models in general.

Using LLMs to solve problems of converting a (not so random) string to OSM opening hours seems like cracking a nut with a sledgehammer. (shooting sparrows with cannons)

I refer to my last earnest post here. What is needed is a model trained on doing this conversion (or actually, a very lenient/flexible parser would do), a KMP library that offers cross-platform OCR and finally a way to run a model in Kotlin/KMP if the conversion is not done by a parser. All of which (if they don't exist yet) should be developed and tested in separate repositories from this one. If that setup works, it should in the end just be a few lines of code to make use of that functionality in this app. (It needn't be a separate app, only if due to license constraints)
